content/codes/components/enigma.ts

```diff
@@ -6,31 +6,598 @@
+		encoding.push(this.reflector.reflect(getLast(encoding)));
+		encoding = this.rotors.reduceRight((acc, r) => [...acc, r.encodeReverse(getLast(acc))], encoding);
+    if (this.skipPlugboard)
+      encoding.push(getLast(encoding));
+    else
+		  encoding.push(this.plugboard.encode(getLast(encoding)));
+
+    if (DEBUG)
+      console.log("Encoding path was: " + encoding.join("->"));
+      
+    return encoding;
+	};
+	
+	toString()
+	{
+ 		var str = this.plugboard.toString();
+ 		str += "\n" + this.rotors.map(r => r.toString()).join("\n");
+		str += "\n" + this.reflector.toString();
+		return str;
+	}
+	
+	rotorPositions()
+	{
+		return this.rotors.map(r => letters[r.getOffset()]);
+	}
+
+	rotate(index:number, up:boolean)
+	{
+		if (up)
+			this.rotors[index].step();
+		else 
+		{
+			for (var i = 0; i < 25; ++i)
+				this.rotors[index].step();
+		}
+	}
+}
+
+// function testEnigma()
+// {
+//   DEBUG = true;
+//   let test = "HELLOWORLDTHISISKENHELLOWORLDTHISISKEN";
+//   let encoding = new Machine(rotorChoices, initialPositions).encodeLetters(test);
+//   console.log(encoding);
+//   let decoding = new Machine(rotorChoices, initialPositions).encodeLetters(encoding);
+//   console.log(decoding);
+//   console.log(test === decoding ? "Matches" : "Doesn't match");
+// }
 
 @register('x-enigma', {templateId: '#enigma'})
 export class Enigma extends CustomElementView {
   private $lights: Obj<ElementView> = {};
 
-  ready() {
+  private rotorChoices = [2, 1, 0];
+  private initialPositions = ['A', 'A', 'A'];
+  
+  private machine:Machine = new Machine(this.rotorChoices, this.initialPositions);
+  public svgview:any;
+
+  ready() { 
     for (const $l of this.$$('.lightboard .key')) {
       this.$lights[$l.text] = $l;
     }
 
     for (const $k of this.$$('.keyboard .key')) {
       const char = $k.text;
-      $k.on('pointerdown', () => this.showLight(this.encode(char)));
+      $k.on('pointerdown', () => {
+        const encodepath:string[] = this.encode(char);
+        this.showLight(getLast(encodepath));
+        this.updateDisplay(encodepath);
+        this.svgview.drawPath(encodepath, false);
+      });
+      $k.on('pointerup', () => {
+        this.hideLight();
+      });
     }
+
+    // callback in case we want to mess with the other things in the gui when the svg keyboard is clicked
+    const outer = this;
+    const keypresscallback = function(encodepath:string[]) {
+      outer.svgview.drawPath(encodepath, true);
+      outer.updateDisplay(encodepath);
+    };
+
+    this.svgview = createEnigmaPathSVG("enigma_svg", this.machine, keypresscallback);
+    this.connectRotorControls();
+    this.svgview.drawPath(null);
+
+    const plugboardview = new PlugboardView("enigmaview", this.machine, this.svgview);
+    document.getElementById('clearciphertext')!.addEventListener("click", () => {
+      document.getElementById('ciphertext')!.innerHTML='&nbsp;';
+      document.getElementById('cipherpath')!.innerHTML='&nbsp;';
```

Comment:
See my comment below: we should never use IDs in element templates, because that breaks if we have multiple instances of the same instance on the page.

Rather than using `getElementById`, you can use Mathigon's built-in DOM element wrappers, e.g. `$()`. And we should only run every selector once, and then save it to a variable, not multiple times. (E.g. in this case, we'd run the same selector over and over again, whenever the click event is triggered.)